### ISSUE TITLE
UnFlow: Unsupervised Learning of Optical Flow with a Bidirectional Census Loss

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#103]
*AAAI 2018*

## Motivation

- Evade the need for hard-to-obtain labeled ground truth data.
- Take occlusion into account for better loss computation.

## Key Idea

- Bidirectional Framework (every component is *symmetric*)
- Model occlusion via forward-backward flow consistency *(for non-occluded pixels, forward and backward flow vectors should be almost identical except for the direction)*
- Incorporate census transform for more robust optical flow constraint (pixel brightness constancy)

## Method

![image](https://user-images.githubusercontent.com/46308407/129853141-fd418b0b-1140-4e0d-9890-5c81d0bd90fc.png)

- Symmetric Framework: (I1, I2) and (I2, I1) are fed to the model
- Enforce 2nd order smoothness to flow map *(only constraint for occluded regions)*
- If two flow maps(forward and backward) differ, those pixels are regarded occluded.
- Additional terms in data loss prevents trivial solution of setting all pixels as *occluded*.
- Census transform is pre-applied for warped images for robust data loss computation. (robust to light changes, etc)

### Occlusion Mask
![image](https://user-images.githubusercontent.com/46308407/129853534-5d8d213a-559f-4ef7-9a3a-1c0e3e765c25.png)
- Note that if a pixel is **not** occluded, the forward and backward flow vector should be opposite signs with identical magnitude.

### Objective
![image](https://user-images.githubusercontent.com/46308407/129853581-ad7116e8-320e-4abd-8062-bae3471ded6d.png)
![image](https://user-images.githubusercontent.com/46308407/129853665-425c058c-8e0a-4742-af1b-bbedd2b8ad26.png)
- Data term enforces photometric consistency after Census Transform
![image](https://user-images.githubusercontent.com/46308407/129853702-1cf1a9bd-4247-4d4b-89fb-2d7152c33105.png)
- Smoothness regularization encourages flow fields to be colinear in the neighbor (locally direction-aligned)
![image](https://user-images.githubusercontent.com/46308407/129853926-9c9e6dd2-3da3-49f1-9e62-c8af810189ef.png)
- For non-occluded pixels, forward and backward flow vectors should be consistent. *(Hence consistency penalty)*

![image](https://user-images.githubusercontent.com/46308407/129854303-29ca0338-33b7-4fbb-887a-3ba9c3edd70a.png)
- Photometric Difference
![image](https://user-images.githubusercontent.com/46308407/129854335-16190a37-e8c2-4b5c-bd6b-e2a2b8decb79.png)
- Robust generalized Charbonnier penalty

### Architecture
- First network is FlowNetC, which explicitly models correlation volume.
- Additionally stack 2 FlowNetS on top of FlowNetC, resulting in UnFlow-CSS for iterative refinement as in *FlowNet2.0*.
![image](https://user-images.githubusercontent.com/46308407/129854160-7cfaf7eb-6ebf-4e03-993f-6d93622749af.png)

### (Optional) Supervised Fine-tuning
- If ground truth optical flow labels are available, one can leverage the information via fine-tuning.
![image](https://user-images.githubusercontent.com/46308407/129854245-6c219a50-6454-4ac4-822e-d95bc51de4ad.png)

### Ablation
![image](https://user-images.githubusercontent.com/46308407/129854451-4e4fa75b-ca59-4939-a4ca-de6552663fe8.png)

### Baseline Comparison
![image](https://user-images.githubusercontent.com/46308407/129854479-fceb8adb-11ea-40cd-9af9-1bf51df3ac77.png)

## Comments

- Symmetric frameworks give the impression of being *elegant*.
- Though not visually well elaborated, the paper is generally a easy read with simple and intuitive structure.
- *"Supervised approaches are often limited by the amount of available ground truth data while unsupervised techniques are limited by the loss function"*
- How faithfully the problem is modeled by the loss is the key.
